src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -3041,6 +3084,74 @@ std::shared_ptr<ParallelReadingExtension> ReadFromMergeTree::getParallelReadingE
         context->getClusterForParallelReplicas()->getShardsInfo().at(0).getAllNodeCount());
 }
 
+void ReadFromMergeTree::replaceColumnsForTextSearch(const IndexReadColumns & added_columns, const Names & removed_columns)
+{
+    if (added_columns.empty())
+        return;
+
+    for (const auto & column_name : removed_columns)
+    {
+        auto it = std::ranges::find(all_column_names, column_name);
+        all_column_names.erase(it);
+    }
+
+    /// We have to recreate virtual columns and storage snapshot to add new virtual columns for reading from text index.
+    auto new_virtual_columns = std::make_shared<VirtualColumnsDescription>(*storage_snapshot->virtual_columns);
+
+    for (const auto & [index_name, columns] : added_columns)
+    {
+        auto [task_it, inserted] = index_read_tasks.try_emplace(index_name);
+        auto & index_task = task_it->second;
+
+        if (inserted)
+        {
+            if (!indexes)
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Index {} not found in analyzed indexes, indexes are not initialized", index_name);
+
+            const auto & useful_indices = indexes->skip_indexes.useful_indices;
+            auto index_it = std::ranges::find_if(useful_indices, [&](const auto & index) { return index.index->index.name == index_name; });
+
+            if (index_it == useful_indices.end())
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Index {} not found in analyzed indexes", index_name);
+
+            index_task.index = *index_it;
+        }
+
+        for (const auto & [column_name, column_type] : columns)
+        {
+            auto it = std::ranges::find(all_column_names, column_name);
+            if (it != all_column_names.end())
+                throw Exception(ErrorCodes::LOGICAL_ERROR, "Column {} already added for reading", column_name);
+
+            all_column_names.push_back(column_name);
+            new_virtual_columns->addEphemeral(column_name, column_type, "");
+            index_task.columns.emplace_back(column_name, column_type);
+        }
+    }
+
+    storage_snapshot = std::make_shared<StorageSnapshot>(
```

Comment:
This will likely break `enable_shared_storage_snapshot_in_query = 1`  https://github.com/ClickHouse/ClickHouse/pull/79471 . I ran into the same issue when developing projection index, and I also haven’t found a good solution yet.

Reply:
You are right